<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключимся-к-базе-данных-и-проимпортируем-библиотеки:" data-toc-modified-id="Подключимся-к-базе-данных-и-проимпортируем-библиотеки:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключимся к базе данных и проимпортируем библиотеки:</a></span></li><li><span><a href="#Изучим-общую-информацию-о-данных" data-toc-modified-id="Изучим-общую-информацию-о-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Изучим общую информацию о данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица <code>books</code></a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица <code>authors</code></a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица <code>publishers</code></a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица <code>ratings</code></a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица <code>reviews</code></a></span></li><li><span><a href="#Подитог:" data-toc-modified-id="Подитог:-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Подитог:</a></span></li></ul></li><li><span><a href="#Выполнение-задач" data-toc-modified-id="Выполнение-задач-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выполнение задач</a></span><ul class="toc-item"><li><span><a href="#Посчитать,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитать,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитать, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитать количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключим-из-анализа-брошюры" data-toc-modified-id="Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключим-из-анализа-брошюры-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры</a></span></li><li><span><a href="#Определить-автора-с-самой-высокой-средней-оценкой-книг-—-учитывать-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определить-автора-с-самой-высокой-средней-оценкой-книг-—-учитывать-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# SQL - анализ база данных сервиса для чтения книг по подписке

# Описание проекта:
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Цели исследования:
Проанализировать информацию о книгах, авторах, рейтингах и данные о пользователских обзорах книг.

# Описание данных
У нас будут данные из следующих источников:

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# ER-диаграмма


# Наши задачи:
- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Подключимся к базе данных и проимпортируем библиотеки:


In [23]:
import pandas as pd 
from sqlalchemy import create_engine

In [24]:
# устанавливаем параметры 
db_config = {'user': 'praktikum_student', # имя пользователя  
             'pwd': 'Sdf4$2;d-d30pp', # пароль  
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',  
             'port': 6432, # порт подключения  
             'db': 'data-analyst-final-project-db'} # название базы данных 
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],  
                                                         db_config['pwd'],  
                                                         db_config['host'],  
                                                         db_config['port'],  
                                                         db_config['db'])
# сохраняем коннектор 
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Изучим общую информацию о данных

In [25]:
# напишем функцию для ознакомления с датафреймами
def read_df(df):
    df.info()
    #display(df.info())
    print("")
    print("Количество дублей:", df.duplicated().sum())
    print("")
    print("Количество пропусков: ")
    display(df.isna().sum())
    display(df.head(3))
    display(df.tail(3))


### Таблица `books`

In [26]:
books = ''' SELECT * FROM books; '''
books = pd.io.sql.read_sql(books, con = engine)

In [27]:
read_df(books)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB

Количество дублей: 0

Количество пропусков: 


book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

book_id  author_id                                              title  \
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143

### Таблица `authors`

In [28]:
authors = ''' SELECT * FROM authors; '''
authors = pd.io.sql.read_sql(authors, con = engine)

In [29]:
read_df(authors)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB

Количество дублей: 0

Количество пропусков: 


author_id    0
author       0
dtype: int64

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie

author_id                                author
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

### Таблица `publishers`

In [30]:
publishers = ''' SELECT * FROM publishers; '''
publishers = pd.io.sql.read_sql(publishers, con = engine)

In [31]:
read_df(publishers)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB

Количество дублей: 0

Количество пропусков: 


publisher_id    0
publisher       0
dtype: int64

publisher_id  publisher
0             1        Ace
1             2   Ace Book
2             3  Ace Books

publisher_id              publisher
337           338  Yale University Press
338           339               Yearling
339           340         Yearling Books

### Таблица `ratings`

In [32]:
ratings = ''' SELECT * FROM ratings; '''
ratings = pd.io.sql.read_sql(ratings, con = engine)

In [33]:
read_df(ratings)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB

Количество дублей: 0

Количество пропусков: 


rating_id    0
book_id      0
username     0
rating       0
dtype: int64

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

rating_id  book_id    username  rating
6453       6454     1000  jarvispaul       5
6454       6455     1000       zross       2
6455       6456     1000     fharris       5

### Таблица `reviews`

In [34]:
reviews = ''' SELECT * FROM reviews; '''
reviews = pd.io.sql.read_sql(reviews, con = engine)

In [35]:
read_df(reviews)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB

Количество дублей: 0

Количество пропусков: 


review_id    0
book_id      0
username     0
text         0
dtype: int64

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

review_id  book_id        username  \
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...

### Подитог:
Данные чистые, пропусков дубликатов не найдено, можно приступать к заданиям.

## Выполнение задач

### Посчитать, сколько книг вышло после 1 января 2000 года

In [36]:
books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [37]:
count_books = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'

'''
pd.io.sql.read_sql(count_books, con = engine)

count
0    819

819 книг вышло после 1го января 2000 года.

### Для каждой книги посчитать количество обзоров и среднюю оценку

In [38]:
count_rating_review = '''
SELECT b.title
       ,COUNT(DISTINCT rv.review_id) AS number_of_ratings
       ,ROUND(AVG(rt.rating), 2) AS avg_rating

FROM books AS b LEFT JOIN reviews AS rv ON b.book_id = rv.book_id
LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
GROUP BY b.book_id

'''
pd.io.sql.read_sql(count_rating_review, con = engine)

title  number_of_ratings  \
0                                         'Salem's Lot                  2   
1                   1 000 Places to See Before You Die                  1   
2    13 Little Blue Envelopes (Little Blue Envelope...                  3   
3    1491: New Revelations of the Americas Before C...                  2   
4                                                 1776                  4   
..                                                 ...                ...   
995           Wyrd Sisters (Discworld  #6; Witches #2)                  3   
996                        Xenocide (Ender's Saga  #3)                  3   
997                                    Year of Wonders                  4   
998                        You Suck (A Love Story  #2)                  2   
999  Zen and the Art of Motorcycle Maintenance: An ...                  4   

     avg_rating  
0          3.67  
1          2.50  
2          4.67  
3          4.50  
4          4.00  
..          ...  
995        3.67  
996        3.40  
997        3.20  
998        4.50  
999        3.83  

[1000 rows x 3 columns]

Посчитано количество обзоров и средняя оценка для каждой книги.

### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры

In [39]:
more_50_pages = '''
SELECT b.publisher_id AS publisher_id
       ,COUNT(b.book_id) AS sum_books
       ,p.publisher AS publisher
FROM books AS b LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE num_pages > 50
GROUP BY b.publisher_id, p.publisher
ORDER BY sum_books DESC
LIMIT 1

'''
pd.io.sql.read_sql(more_50_pages, con = engine)

publisher_id  sum_books      publisher
0           212         42  Penguin Books

Издательство `Penguin Books` - является издательством, что выпустила наибольшее число книг с количеством страниц больше 50.

### Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré со средним рейтингом 4.29.

In [79]:
top_author = '''
SELECT  aut.author
        ,ROUND(AVG(r.rating), 2) AS avg_rating
FROM (SELECT book_id
             ,AVG(rating) AS rating_avg
      FROM ratings
      GROUP BY book_id
      HAVING COUNT(rating_id) >= 50) AS rt
LEFT JOIN ratings AS r on rt.book_id = r.book_id 
LEFT JOIN books AS b ON rt.book_id = b.book_id
LEFT JOIN authors AS aut ON b.author_id = aut.author_id
GROUP BY aut.author       
ORDER BY avg_rating DESC
LIMIT 5
'''
pd.io.sql.read_sql(top_author, con = engine)

author  avg_rating
0         J.K. Rowling/Mary GrandPré        4.29
1  Markus Zusak/Cao Xuân Việt Khương        4.26
2                     J.R.R. Tolkien        4.25
3                  Louisa May Alcott        4.19
4                       Rick Riordan        4.08

### Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [41]:
avg_reviews = '''
SELECT COUNT(DISTINCT rev.review_id)/COUNT(DISTINCT rev.username) as avg_reviews
FROM ratings LEFT JOIN reviews as rev on ratings.username = rev.username
WHERE ratings.username in (SELECT username
                       FROM ratings
                       GROUP BY username
                       HAVING COUNT(rating_id) > 50)
'''
pd.io.sql.read_sql(avg_reviews, con = engine)

avg_reviews
0           24

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет 24.

## Вывод

- 819 книг вышло после 1го января 2000 года.
- Посчитано количество обзоров и средняя оценка для каждой книги.
- Издательство Penguin Books - является издательством, что выпустила наибольшее число книг с количеством страниц больше 50.
- Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré со средним рейтингом 4.29.
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет 24.